### ISSUE TITLE
SSH key exchanges fail unexpectedly with "unable to write X bytes of raw data" errors due to small ProFTPD buffer

### ISSUE THREAD

**Issue by jlecour** [proftpd/proftpd#1694]
### What I Did

After upgrading from Debian 11 (ProFTPd 1.3.7) to Debian 12 (ProFTPd 1.3.8) I can't connect using SFTP.

In server logs I have this :

```
2023-08-05 18:57:02,433 mod_sftp/1.1.1[581227]: client (<REDACTED_IP>:51472) connected to server (<REDACTED_IP>:2022) [session ID ZM5/Xh+qC30fqgmBAAjea6TB]
2023-08-05 18:57:02,434 mod_sftp/1.1.1[581227]: sent server version 'SSH-2.0-FTP Server Ready'
2023-08-05 18:57:02,435 mod_sftp/1.1.1[581227]: received client version 'SSH-2.0-Cyberduck/8.4.1.38006 (Mac OS X/12.6.7) (x86_64)'
2023-08-05 18:57:02,435 mod_sftp/1.1.1[581227]: handling connection from SSH2 client 'Cyberduck/8.4.1.38006'
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session key exchange: curve25519-sha256
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session host key algorithm: ecdsa-sha2-nistp256
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session server hostkey: ecdsa-sha2-nistp256
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session client-to-server encryption: aes128-cbc
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session server-to-client encryption: aes128-cbc
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session client-to-server MAC: hmac-sha1
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session server-to-client MAC: hmac-sha1
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session client-to-server compression: none
2023-08-05 18:57:02,478 mod_sftp/1.1.1[581227]:  + Session server-to-client compression: none
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: message format error: unable to write 104 bytes of raw data (buflen = 96)
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: -----BEGIN STACK TRACE-----
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [0] /usr/lib/proftpd/mod_sftp.so(sftp_msg_write_data+0x92) [0x7f7675c4aa62]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [1] /usr/lib/proftpd/mod_sftp.so(+0x23c39) [0x7f7675c55c39]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [2] /usr/lib/proftpd/mod_sftp.so(sftp_kex_handle+0x1907) [0x7f7675c5b497]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [3] /usr/lib/proftpd/mod_sftp.so(+0x1b93c) [0x7f7675c4d93c]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [4] /usr/lib/proftpd/mod_sftp.so(sftp_ssh2_packet_process+0x51) [0x7f7675c4dcb1]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [5] /usr/lib/proftpd/mod_sftp.so(+0x155f6) [0x7f7675c475f6]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [6] proftpd: (authenticating) - (connecting): IDLE(+0x205bf) [0x55e28e4e45bf]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [7] proftpd: (authenticating) - (connecting): IDLE(+0x20ea1) [0x55e28e4e4ea1]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [8] proftpd: (authenticating) - (connecting): IDLE(main+0x5f8) [0x55e28e4e2b98]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [9] /lib/x86_64-linux-gnu/libc.so.6(+0x271ca) [0x7f76768461ca]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [10] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x85) [0x7f7676846285]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: [11] proftpd: (authenticating) - (connecting): IDLE(_start+0x21) [0x55e28e4e31c1]
2023-08-05 18:57:02,515 mod_sftp/1.1.1[581227]: -----END STACK TRACE-----
```

### What I Expected/Wanted

I wanted to be able to connect over SFTP.

### ProFTPD Version and Configuration

```bash
# proftpd -V
Compile-time Settings:
  Version: 1.3.8 (stable)
  Platform: LINUX [Linux 6.1.0-10-cloud-amd64 x86_64]
  OS/Release:
    PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
    NAME="Debian GNU/Linux"
    VERSION_ID="12"
    VERSION="12 (bookworm)"
    VERSION_CODENAME=bookworm
    ID=debian
  Built: Wed Jun 28 2023 13:17:54 UTC
  Built With:
    configure  '--infodir=/share/info' '--disable-option-checking' '--disable-silent-rules' '--libdir=/lib/x86_64-linux-gnu' '--disable-dependency-tracking' '--prefix=/usr' '--with-pkgconfig=lib/pkgconfig' '--with-includes=/usr/include/postgresql:/usr/include/mariadb:/usr/include/mariadb/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-openssl' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre2' '--disable-strip' '--enable-redis' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp:mod_digest:mod_ident:mod_log_forensic:mod_qos:mod_statcache:mod_tls_fscache:mod_tls_shmcache:mod_dnsbl' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -ffile-prefix-map=/build/proftpd-dfsg-NLy2xM/proftpd-dfsg-1.3.8+dfsg=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-z,relro -Wl,-z,now' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -ffile-prefix-map=/build/proftpd-dfsg-NLy2xM/proftpd-dfsg-1.3.8+dfsg=. -fstack-protector-strong -Wformat -Werror=format-security'

  CFLAGS: -g2 -g -O2 -ffile-prefix-map=/build/proftpd-dfsg-NLy2xM/proftpd-dfsg-1.3.8+dfsg=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -fno-strict-aliasing -Werror=implicit-function-declaration
  LDFLAGS: -Wl,-L$(top_srcdir)/lib,-L$(top_builddir)/lib -Wl,-z,relro -Wl,-z,now -rdynamic  -L/usr/lib/x86_64-linux-gnu/ -L/usr/lib/x86_64-linux-gnu
  LIBS: -lacl  -lpcre2-posix -lpcre2-8 -lssl -lcrypto -lsodium -lcap  -lpam -lattr -lidn2 -lnsl -lresolv -lresolv -lcrypt -lhiredis -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    + ncursesw support
    + NLS support
    + OpenSSL support (OpenSSL 3.0.9 30 May 2023)
    - PCRE support
    + PCRE2 support
    + POSIX ACL support
    + Redis support
    + Sendfile support
    + Shadow file support
    + Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

```
# dpkg -l proftpd-*
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name               Version              Architecture Description
+++-==================-====================-============-============================================================
un  proftpd-abi-1.3.8  <none>               <none>       (no description available)
ii  proftpd-basic      1.3.8+dfsg-4+deb12u1 all          Transitional dummy package for ProFTPD
ii  proftpd-core       1.3.8+dfsg-4+deb12u1 amd64        Versatile, virtual-hosting FTP daemon - binaries
un  proftpd-dev        <none>               <none>       (no description available)
un  proftpd-doc        <none>               <none>       (no description available)
ii  proftpd-mod-crypto 1.3.8+dfsg-4+deb12u1 amd64        Versatile, virtual-hosting FTP daemon - TLS/SSL/SFTP modules
un  proftpd-mod-dnsbl  <none>               <none>       (no description available)
un  proftpd-mod-geoip  <none>               <none>       (no description available)
un  proftpd-mod-ldap   <none>               <none>       (no description available)
un  proftpd-mod-mysql  <none>               <none>       (no description available)
un  proftpd-mod-odbc   <none>               <none>       (no description available)
un  proftpd-mod-pgsql  <none>               <none>       (no description available)
un  proftpd-mod-snmp   <none>               <none>       (no description available)
un  proftpd-mod-sqlite <none>               <none>       (no description available)
ii  proftpd-mod-wrap   1.3.8+dfsg-4+deb12u1 amd64        Versatile, virtual-hosting FTP daemon - tcpwrapper module
```

```
# cat /etc/proftpd/proftpd.conf
#
# /etc/proftpd/proftpd.conf -- This is a basic ProFTPD configuration file.
# To really apply changes, reload proftpd after modifications, if
# it runs in daemon mode. It is not required in inetd/xinetd mode.
# 

# Includes DSO modules
Include /etc/proftpd/modules.conf

# Set off to disable IPv6 support which is annoying on IPv4 only boxes.
UseIPv6				on
# If set on you can experience a longer connection delay in many cases.
<IfModule mod_ident.c>
IdentLookups			off
</IfModule>


ServerName			"Debian"
# Set to inetd only if you would run proftpd by inetd/xinetd.
# Read README.Debian for more information on proper configuration.
ServerType				standalone
DeferWelcome			off

MultilineRFC2228		on
DefaultServer			on
ShowSymlinks			on

TimeoutNoTransfer		600
TimeoutStalled			600
TimeoutIdle			1200

DisplayLogin                    welcome.msg
DisplayChdir               	.message true
ListOptions                	"-l"

DenyFilter			\*.*/

# Use this to jail all users in their homes 
# DefaultRoot			~

# Users require a valid shell listed in /etc/shells to login.
# Use this directive to release that constrain.
# RequireValidShell		off

# Port 21 is the standard FTP port.
#Port				21

# In some cases you have to specify passive ports range to by-pass
# firewall limitations. Ephemeral ports can be used for that, but
# feel free to use a more narrow range.
# PassivePorts                  49152 65534

# If your host was NATted, this option is useful in order to
# allow passive tranfers to work. You have to use your public
# address and opening the passive ports used on your firewall as well.
# MasqueradeAddress		1.2.3.4

# This is useful for masquerading address with dynamic IPs:
# refresh any configured MasqueradeAddress directives every 8 hours
<IfModule mod_dynmasq.c>
# DynMasqRefresh 28800
</IfModule>

# To prevent DoS attacks, set the maximum number of child processes
# to 30.  If you need to allow more than 30 concurrent connections
# at once, simply increase this value.  Note that this ONLY works
# in standalone mode, in inetd mode you should use an inetd server
# that allows you to limit maximum number of processes per service
# (such as xinetd)
MaxInstances			30

# Set the user and group that the server normally runs at.
User				proftpd
Group				nogroup

# Umask 022 is a good standard umask to prevent new files and dirs
# (second parm) from being group and world writable.
Umask				022  022
# Normally, we want files to be overwriteable.
AllowOverwrite			on

# Uncomment this if you are using NIS or LDAP via NSS to retrieve passwords:
# PersistentPasswd		off

# This is required to use both PAM-based authentication and local passwords
# AuthOrder			mod_auth_pam.c* mod_auth_unix.c

# Be warned: use of this directive impacts CPU average load!
# Uncomment this if you like to see progress and transfer rate with ftpwho
# in downloads. That is not needed for uploads rates.
#
# UseSendFile			off

TransferLog /var/log/proftpd/xferlog
SystemLog   /var/log/proftpd/proftpd.log

# Logging onto /var/log/lastlog is enabled but set to off by default
#UseLastlog on

# In order to keep log file dates consistent after chroot, use timezone info
# from /etc/localtime.  If this is not set, and proftpd is configured to
# chroot (e.g. DefaultRoot or <Anonymous>), it will use the non-daylight
# savings timezone regardless of whether DST is in effect.
#SetEnv TZ :/etc/localtime

<IfModule mod_quotatab.c>
QuotaEngine off
</IfModule>

<IfModule mod_ratio.c>
Ratios off
</IfModule>


# Delay engine reduces impact of the so-called Timing Attack described in
# http://www.securityfocus.com/bid/11430/discuss
# It is on by default. 
<IfModule mod_delay.c>
DelayEngine on
</IfModule>

<IfModule mod_ctrls.c>
ControlsEngine        off
ControlsMaxClients    2
ControlsLog           /var/log/proftpd/controls.log
ControlsInterval      5
ControlsSocket        /var/run/proftpd/proftpd.sock
</IfModule>

<IfModule mod_ctrls_admin.c>
AdminControlsEngine off
</IfModule>

#
# Alternative authentication frameworks
#
#Include /etc/proftpd/ldap.conf
#Include /etc/proftpd/sql.conf

#
# This is used for FTPS connections
#
#Include /etc/proftpd/tls.conf

#
# Useful to keep VirtualHost/VirtualRoot directives separated
#
#Include /etc/proftpd/virtuals.conf

# A basic anonymous configuration, no upload directories.

# <Anonymous ~ftp>
#   User				ftp
#   Group				nogroup
#   # We want clients to be able to login with "anonymous" as well as "ftp"
#   UserAlias			anonymous ftp
#   # Cosmetic changes, all files belongs to ftp user
#   DirFakeUser	on ftp
#   DirFakeGroup on ftp
# 
#   RequireValidShell		off
# 
#   # Limit the maximum number of anonymous logins
#   MaxClients			10
# 
#   # We want 'welcome.msg' displayed at login, and '.message' displayed
#   # in each newly chdired directory.
#   DisplayLogin			welcome.msg
#   DisplayChdir		.message
# 
#   # Limit WRITE everywhere in the anonymous chroot
#   <Directory *>
#     <Limit WRITE>
#       DenyAll
#     </Limit>
#   </Directory>
# 
#   # Uncomment this if you're brave.
#   # <Directory incoming>
#   #   # Umask 022 is a good standard umask to prevent new files and dirs
#   #   # (second parm) from being group and world writable.
#   #   Umask				022  022
#   #            <Limit READ WRITE>
#   #            DenyAll
#   #            </Limit>
#   #            <Limit STOR>
#   #            AllowAll
#   #            </Limit>
#   # </Directory>
# 
# </Anonymous>

# Include other custom configuration files
Include /etc/proftpd/conf.d/
```

```
# cat /etc/proftpd/conf.d/custom.conf 

ServerName        "My FTP Server"
ServerIdent    on "FTP Server Ready"
AccessGrantMsg    "Hey, bienvenue %u sur le serveur FTP !"
AuthUserFile      /etc/proftpd/vpasswd

# Connexion limits
MaxInstances      50
MaxClients        40
MaxClientsPerHost 20

# Network settings
PassivePorts  60000 61000
UseReverseDNS off
IdentLookups  off
TimesGMT      off
<IfModule !mod_sftp.c>
Port 21
</IfModule>

# Local permissions
DefaultRoot       ~
Umask             137  027
RequireValidShell off
UseFtpUsers       off

# Allow RESUME (REST command)
AllowStoreRestart on

AuthOrder    mod_auth_file.c
AuthUserFile /etc/proftpd/vpasswd

<IfModule mod_sftp.c>
    SFTPEngine   on
    Port         2022
    #DefaultRoot  ~
    
    SFTPLog      /var/log/proftpd/sftp.log
    
    SFTPAuthMethods password publickey
    # ProFTPD doesn't support ed25519 yet
    # SFTPHostKey /etc/ssh/ssh_host_ed25519_key
    SFTPHostKey /etc/ssh/ssh_host_ecdsa_key
    SFTPHostKey /etc/ssh/ssh_host_rsa_key
    
    SFTPAuthorizedUserKeys file:/etc/proftpd/sftp.passwd.keys/%u
    AuthUserFile /etc/proftpd/vpasswd
    
    RequireValidShell off

    # Enable compression
    # SFTPCompression delayed
</IfModule>
```

Thanks for maintaining ProFTPd and for any help I might get here.

**Comment by Castaglia** [comment]
Does this error occur every time you connect with this Cyberduck client, or only sometimes?  Do you see this error with other SFTP clients?

**Comment by jlecour** [comment]
Hi,

I've just tried with FileZilla on a Debian 12 laptop and it works.
I'll continue my differential diagnosis and will report my findings.

Thanks.

**Comment by Castaglia** [comment]
I also tried using the [Cyberduck CLI](https://duck.sh/) to reproduce this locally, but it succeeded:
```
2023-08-06 21:18:02,173 mod_sftp/1.1.1[14319]: received client version 'SSH-2.0-Cyberduck/8.6.2.40032 (Linux/4.9.49-moby) (amd64)'
```
Perhaps it's something related to your specific ECDSA 256 hostkey?

The `mod_sftp` module _does_ use some fixed-size buffers for portions of the SFTP key exchange ("KEX") sequence, and the error you reported looks like one of those buffers is too small.  Unfortunately, the module is _not_ great at telling us which particular buffer is too small.  Once I can reproduce this issue locally, then increasing the size of the correct buffer should be quick.

**Comment by ylangisc** [comment]
Possibly related to https://github.com/proftpd/proftpd/issues/1595 as you use Cyberduck 8.4.1 which does still prefer `aes128-cbc`. Using the latest version should resolve the issue. Nonetheless it looks like a server side issue.

**Comment by jlecour** [comment]
> Using the latest version should resolve the issue.

You mean the latest version of Cyberduck? I've updated this morning and I still had the problem.

**Comment by Castaglia** [comment]
> I've tried every combination of host keys at hand and nothing works :/
> 
> Any idea how I could help debugging this?

Could you send me the hostkey files you are using, via direct email?

In order to address this, I need to be able to reproduce the issue locally.  It is possible that the issues you are encountering are related to the specific hostkeys you have generated.  Maybe something about them, _vs_ (say) new/different hostkeys, is using more of the fixed-size buffers than expected?

Another thing that would help would be to get more detailed logging, using trace logging, _e.g._:
```
TraceLog /path/to/proftpd-sftp.log
Trace ssh2:30 sftp:30
```
The trace logging might help us narrow down just where, during the SSH key exchange protocol, this error is occurring.

**Comment by ylangisc** [comment]
> > Using the latest version should resolve the issue.
> 
> You mean the latest version of Cyberduck? I've updated this morning and I still had the problem.

Yes. Looking at your initial report again it looks like the KEX init was successful (with `aes128-cbc`) but failing in the subsequent authentication phase. Please ignore my comment.

**Comment by ylangisc** [comment]
From the logs I see that the issue is with the `SSH_MSG_KEXDH_INIT` message which basically contains a session based generated EC public key of 32 bytes. Wonder why the package size is not static. Would be interesting how a wireshark debug would look like. On my setup (works fine) it looks like:
<img width="1568" alt="ssh-wireshark" src="https://github.com/proftpd/proftpd/assets/655240/29fbe596-12b2-4ca9-b51a-195737f2dc58">

**Comment by Castaglia** [comment]
@jlecour I don't know you're able to compile ProFTPD from source, and thus test out a patch.  If you are, https://github.com/proftpd/proftpd/pull/1704 might fix this.

**Comment by jlecour** [comment]
> Could you send me the hostkey files you are using, via direct email?

I've generated a brand new set of host keys with `ssh-keygen` and the result is the same. If you still want that keys, I can share them and make some new ones again for me.

**Comment by jlecour** [comment]
I have managed to get the source and compile them, run the tests…
But my SFTP configuration is ignored. I figured that I had not enabled the module.

So I've compiled again with `mod_sftp` and had the same result.

I've tried with "mod_sftp" and "mod_sftp_pam" but then I get compile errors likethis :

```
cd modules/ && make static
make[1]: Entering directory '/root/proftpd/modules'
gcc -DHAVE_CONFIG_H  -DLINUX  -I.. -I../include -I../include  -g2 -O2 -Wall -fno-omit-frame-pointer -fno-strict-aliasing -c mod_sftp_pam.c
mod_sftp_pam.c:37:3: error: #error "mod_sftp_pam requires PAM support on your system"
   37 | # error "mod_sftp_pam requires PAM support on your system"
      |   ^~~~~
mod_sftp_pam.c:101:8: error: unknown type name ‘pam_handle_t’
  101 | static pam_handle_t *sftppam_pamh = NULL;
      |        ^~~~~~~~~~~~
mod_sftp_pam.c:118:12: warning: ‘struct pam_response’ declared inside parameter list will not be visible outside of this definition or declaration
  118 |     struct pam_response **resps, void *app_data) {
      |            ^~~~~~~~~~~~
mod_sftp_pam.c:117:60: warning: ‘struct pam_message’ declared inside parameter list will not be visible outside of this definition or declaration
  117 | static int sftppam_converse(int nmsgs, PR_PAM_CONST struct pam_message **msgs,
      |                                                            ^~~~~~~~~~~
mod_sftp_pam.c: In function ‘sftppam_converse’:
mod_sftp_pam.c:126:15: error: ‘PAM_MAX_NUM_MSG’ undeclared (first use in this function)
  126 |       nmsgs > PAM_MAX_NUM_MSG) {
      |               ^~~~~~~~~~~~~~~
mod_sftp_pam.c:126:15: note: each undeclared identifier is reported only once for each function it appears in
mod_sftp_pam.c:128:12: error: ‘PAM_CONV_ERR’ undeclared (first use in this function)
  128 |     return PAM_CONV_ERR;
      |            ^~~~~~~~~~~~
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:144:9: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  144 |     if (SFTP_PAM_MSG_MEMBER(msgs, i, msg_style) == PAM_TEXT_INFO) {
      |         ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:144:52: error: ‘PAM_TEXT_INFO’ undeclared (first use in this function)
  144 |     if (SFTP_PAM_MSG_MEMBER(msgs, i, msg_style) == PAM_TEXT_INFO) {
      |                                                    ^~~~~~~~~~~~~
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:148:11: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  148 |           SFTP_PAM_MSG_MEMBER(msgs, i, msg));
      |           ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:152:11: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  152 |           SFTP_PAM_MSG_MEMBER(msgs, i, msg));
      |           ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:154:31: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  154 |         sftp_auth_send_banner(SFTP_PAM_MSG_MEMBER(msgs, i, msg));
      |                               ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:176:16: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  176 |     } else if (SFTP_PAM_MSG_MEMBER(msgs, i, msg_style) == PAM_ERROR_MSG) {
      |                ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:176:59: error: ‘PAM_ERROR_MSG’ undeclared (first use in this function); did you mean ‘PR_ERROR_MSG’?
  176 |     } else if (SFTP_PAM_MSG_MEMBER(msgs, i, msg_style) == PAM_ERROR_MSG) {
      |                                                           ^~~~~~~~~~~~~
      |                                                           PR_ERROR_MSG
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:178:40: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  178 |         "received PAM_ERROR_MSG '%s'", SFTP_PAM_MSG_MEMBER(msgs, i, msg));
      |                                        ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:184:7: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  184 |       SFTP_PAM_MSG_MEMBER(msgs, i, msg));
      |       ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:190:12: error: ‘PAM_SUCCESS’ undeclared (first use in this function); did you mean ‘EXIT_SUCCESS’?
  190 |     return PAM_SUCCESS;
      |            ^~~~~~~~~~~
      |            EXIT_SUCCESS
mod_sftp_pam.c:206:30: error: invalid application of ‘sizeof’ to incomplete type ‘struct pam_response’
  206 |   res = calloc(nmsgs, sizeof(struct pam_response));
      |                              ^~~~~~
mod_sftp_pam.c:209:12: error: ‘PAM_BUF_ERR’ undeclared (first use in this function)
  209 |     return PAM_BUF_ERR;
      |            ^~~~~~~~~~~
mod_sftp_pam.c:213:8: error: invalid use of undefined type ‘struct pam_response’
  213 |     res[i].resp_retcode = 0;
      |        ^
mod_sftp_pam.c:213:11: error: invalid use of undefined type ‘struct pam_response’
  213 |     res[i].resp_retcode = 0;
      |           ^
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:215:13: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  215 |     switch (SFTP_PAM_MSG_MEMBER(msgs, i, msg_style)) {
      |             ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:216:12: error: ‘PAM_PROMPT_ECHO_ON’ undeclared (first use in this function)
  216 |       case PAM_PROMPT_ECHO_ON:
      |            ^~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:219:11: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  219 |           SFTP_PAM_MSG_MEMBER(msgs, i, msg), recvd_responses[i]);
      |           ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:220:12: error: invalid use of undefined type ‘struct pam_response’
  220 |         res[i].resp = strdup(recvd_responses[i]);
      |            ^
mod_sftp_pam.c:220:15: error: invalid use of undefined type ‘struct pam_response’
  220 |         res[i].resp = strdup(recvd_responses[i]);
      |               ^
mod_sftp_pam.c:223:12: error: ‘PAM_PROMPT_ECHO_OFF’ undeclared (first use in this function)
  223 |       case PAM_PROMPT_ECHO_OFF:
      |            ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:226:11: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  226 |           SFTP_PAM_MSG_MEMBER(msgs, i, msg));
      |           ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:227:12: error: invalid use of undefined type ‘struct pam_response’
  227 |         res[i].resp = strdup(recvd_responses[i]);
      |            ^
mod_sftp_pam.c:227:15: error: invalid use of undefined type ‘struct pam_response’
  227 |         res[i].resp = strdup(recvd_responses[i]);
      |               ^
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:232:11: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  232 |           SFTP_PAM_MSG_MEMBER(msgs, i, msg));
      |           ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:233:12: error: invalid use of undefined type ‘struct pam_response’
  233 |         res[i].resp = NULL;
      |            ^
mod_sftp_pam.c:233:15: error: invalid use of undefined type ‘struct pam_response’
  233 |         res[i].resp = NULL;
      |               ^
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:238:11: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  238 |           SFTP_PAM_MSG_MEMBER(msgs, i, msg));
      |           ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:239:12: error: invalid use of undefined type ‘struct pam_response’
  239 |         res[i].resp = NULL;
      |            ^
mod_sftp_pam.c:239:15: error: invalid use of undefined type ‘struct pam_response’
  239 |         res[i].resp = NULL;
      |               ^
mod_sftp_pam.c:76:60: error: invalid use of undefined type ‘const struct pam_message’
   76 | # define SFTP_PAM_MSG_MEMBER(msg, n, member)    ((msg)[(n)]->member)
      |                                                            ^~
mod_sftp_pam.c:253:11: note: in expansion of macro ‘SFTP_PAM_MSG_MEMBER’
  253 |           SFTP_PAM_MSG_MEMBER(msgs, i, msg_style));
      |           ^~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:255:18: error: invalid use of undefined type ‘struct pam_response’
  255 |           if (res[i].resp != NULL) {
      |                  ^
mod_sftp_pam.c:255:21: error: invalid use of undefined type ‘struct pam_response’
  255 |           if (res[i].resp != NULL) {
      |                     ^
mod_sftp_pam.c:256:21: error: invalid use of undefined type ‘struct pam_response’
  256 |             free(res[i].resp);
      |                     ^
mod_sftp_pam.c:256:24: error: invalid use of undefined type ‘struct pam_response’
  256 |             free(res[i].resp);
      |                        ^
mod_sftp_pam.c:257:16: error: invalid use of undefined type ‘struct pam_response’
  257 |             res[i].resp = NULL;
      |                ^
mod_sftp_pam.c:257:19: error: invalid use of undefined type ‘struct pam_response’
  257 |             res[i].resp = NULL;
      |                   ^
mod_sftp_pam.c: At top level:
mod_sftp_pam.c:271:21: error: variable ‘sftppam_conv’ has initializer but incomplete type
  271 | static const struct pam_conv sftppam_conv = { &sftppam_converse, NULL };
      |                     ^~~~~~~~
mod_sftp_pam.c:271:47: warning: excess elements in struct initializer
  271 | static const struct pam_conv sftppam_conv = { &sftppam_converse, NULL };
      |                                               ^
mod_sftp_pam.c:271:47: note: (near initialization for ‘sftppam_conv’)
mod_sftp_pam.c:271:66: warning: excess elements in struct initializer
  271 | static const struct pam_conv sftppam_conv = { &sftppam_converse, NULL };
      |                                                                  ^~~~
mod_sftp_pam.c:271:66: note: (near initialization for ‘sftppam_conv’)
mod_sftp_pam.c: In function ‘sftppam_driver_open’:
mod_sftp_pam.c:299:26: error: ‘PAM_MAX_MSG_SIZE’ undeclared (first use in this function); did you mean ‘EVP_MAX_MD_SIZE’?
  299 |   if (sftppam_userlen > (PAM_MAX_MSG_SIZE + 1)) {
      |                          ^~~~~~~~~~~~~~~~
      |                          EVP_MAX_MD_SIZE
mod_sftp_pam.c:352:9: warning: implicit declaration of function ‘pam_start’; did you mean ‘va_start’? [-Wimplicit-function-declaration]
  352 |   res = pam_start(sftppam_service, sftppam_user, &sftppam_conv, &sftppam_pamh);
      |         ^~~~~~~~~
      |         va_start
mod_sftp_pam.c:353:14: error: ‘PAM_SUCCESS’ undeclared (first use in this function); did you mean ‘EXIT_SUCCESS’?
  353 |   if (res != PAM_SUCCESS) {
      |              ^~~~~~~~~~~
      |              EXIT_SUCCESS
mod_sftp_pam.c:362:12: error: ‘PAM_SYSTEM_ERR’ undeclared (first use in this function); did you mean ‘ERR_SYSTEM_ERROR’?
  362 |       case PAM_SYSTEM_ERR:
      |            ^~~~~~~~~~~~~~
      |            ERR_SYSTEM_ERROR
mod_sftp_pam.c:367:12: error: ‘PAM_BUF_ERR’ undeclared (first use in this function)
  367 |       case PAM_BUF_ERR:
      |            ^~~~~~~~~~~
mod_sftp_pam.c:376:3: warning: implicit declaration of function ‘pam_set_item’ [-Wimplicit-function-declaration]
  376 |   pam_set_item(sftppam_pamh, PAM_RUSER, sftppam_user);
      |   ^~~~~~~~~~~~
mod_sftp_pam.c:376:30: error: ‘PAM_RUSER’ undeclared (first use in this function)
  376 |   pam_set_item(sftppam_pamh, PAM_RUSER, sftppam_user);
      |                              ^~~~~~~~~
mod_sftp_pam.c:377:30: error: ‘PAM_RHOST’ undeclared (first use in this function)
  377 |   pam_set_item(sftppam_pamh, PAM_RHOST, session.c->remote_name);
      |                              ^~~~~~~~~
mod_sftp_pam.c:386:32: error: ‘PAM_TTY’ undeclared (first use in this function); did you mean ‘N_TTY’?
  386 |     pam_set_item(sftppam_pamh, PAM_TTY, sftppam_tty);
      |                                ^~~~~~~
      |                                N_TTY
mod_sftp_pam.c: In function ‘sftppam_driver_authenticate’:
mod_sftp_pam.c:431:9: warning: implicit declaration of function ‘pam_authenticate’; did you mean ‘pr_auth_authenticate’? [-Wimplicit-function-declaration]
  431 |   res = pam_authenticate(sftppam_pamh, 0);
      |         ^~~~~~~~~~~~~~~~
      |         pr_auth_authenticate
mod_sftp_pam.c:432:14: error: ‘PAM_SUCCESS’ undeclared (first use in this function); did you mean ‘EXIT_SUCCESS’?
  432 |   if (res != PAM_SUCCESS) {
      |              ^~~~~~~~~~~
      |              EXIT_SUCCESS
mod_sftp_pam.c:434:12: error: ‘PAM_USER_UNKNOWN’ undeclared (first use in this function); did you mean ‘B_ASN1_UNKNOWN’?
  434 |       case PAM_USER_UNKNOWN:
      |            ^~~~~~~~~~~~~~~~
      |            B_ASN1_UNKNOWN
mod_sftp_pam.c:444:7: warning: implicit declaration of function ‘pam_strerror’; did you mean ‘gai_strerror’? [-Wimplicit-function-declaration]
  444 |       pam_strerror(sftppam_pamh, res));
      |       ^~~~~~~~~~~~
      |       gai_strerror
mod_sftp_pam.c:443:54: warning: format ‘%s’ expects argument of type ‘char *’, but argument 6 has type ‘int’ [-Wformat=]
  443 |       "PAM authentication error (%d) for user '%s': %s", res, user,
      |                                                     ~^
      |                                                      |
      |                                                      char *
      |                                                     %d
  444 |       pam_strerror(sftppam_pamh, res));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                 
      |       |
      |       int
mod_sftp_pam.c:40:41: warning: format ‘%s’ expects argument of type ‘char *’, but argument 5 has type ‘int’ [-Wformat=]
   40 | #define MOD_SFTP_PAM_VERSION            "mod_sftp_pam/0.3"
      |                                         ^~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:445:38: note: in expansion of macro ‘MOD_SFTP_PAM_VERSION’
  445 |     (void) pr_log_pri(PR_LOG_NOTICE, MOD_SFTP_PAM_VERSION
      |                                      ^~~~~~~~~~~~~~~~~~~~
mod_sftp_pam.c:446:56: note: format string is defined here
  446 |       ": PAM authentication error (%d) for user '%s': %s", res, user,
      |                                                       ~^
      |                                                        |
      |                                                        char *
      |                                                       %d
mod_sftp_pam.c:456:9: warning: implicit declaration of function ‘pam_acct_mgmt’ [-Wimplicit-function-declaration]
  456 |   res = pam_acct_mgmt(sftppam_pamh, 0);
      |         ^~~~~~~~~~~~~
mod_sftp_pam.c:466:12: error: ‘PAM_ACCT_EXPIRED’ undeclared (first use in this function)
  466 |       case PAM_ACCT_EXPIRED:
      |            ^~~~~~~~~~~~~~~~
mod_sftp_pam.c:492:52: warning: format ‘%s’ expects argument of type ‘char *’, but argument 6 has type ‘int’ [-Wformat=]
  492 |       "PAM account mgmt error (%d) for user '%s': %s", res, user,
      |                                                   ~^
      |                                                    |
      |                                                    char *
      |                                                   %d
  493 |       pam_strerror(sftppam_pamh, res));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~               
      |       |
      |       int
mod_sftp_pam.c:502:9: warning: implicit declaration of function ‘pam_open_session’ [-Wimplicit-function-declaration]
  502 |   res = pam_open_session(sftppam_pamh, 0);
      |         ^~~~~~~~~~~~~~~~
mod_sftp_pam.c:507:47: warning: format ‘%s’ expects argument of type ‘char *’, but argument 6 has type ‘int’ [-Wformat=]
  507 |       "PAM session error (%d) for user '%s': %s", res, user,
      |                                              ~^
      |                                               |
      |                                               char *
      |                                              %d
  508 |       pam_strerror(sftppam_pamh, res));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~          
      |       |
      |       int
mod_sftp_pam.c:520:9: warning: implicit declaration of function ‘pam_setcred’ [-Wimplicit-function-declaration]
  520 |   res = pam_setcred(sftppam_pamh, PAM_ESTABLISH_CRED);
      |         ^~~~~~~~~~~
mod_sftp_pam.c:520:35: error: ‘PAM_ESTABLISH_CRED’ undeclared (first use in this function); did you mean ‘TCP_ESTABLISHED’?
  520 |   res = pam_setcred(sftppam_pamh, PAM_ESTABLISH_CRED);
      |                                   ^~~~~~~~~~~~~~~~~~
      |                                   TCP_ESTABLISHED
mod_sftp_pam.c:524:12: error: ‘PAM_CRED_EXPIRED’ undeclared (first use in this function)
  524 |       case PAM_CRED_EXPIRED:
      |            ^~~~~~~~~~~~~~~~
mod_sftp_pam.c:542:51: warning: format ‘%s’ expects argument of type ‘char *’, but argument 6 has type ‘int’ [-Wformat=]
  542 |       "PAM credentials error (%d) for user '%s': %s", res, user,
      |                                                  ~^
      |                                                   |
      |                                                   char *
      |                                                  %d
  543 |       pam_strerror(sftppam_pamh, res));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~              
      |       |
      |       int
mod_sftp_pam.c: In function ‘sftppam_exit_ev’:
mod_sftp_pam.c:683:37: error: ‘PAM_DELETE_CRED’ undeclared (first use in this function)
  683 |     res = pam_setcred(sftppam_pamh, PAM_DELETE_CRED);
      |                                     ^~~~~~~~~~~~~~~
mod_sftp_pam.c:685:16: error: ‘PAM_SUCCESS’ undeclared (first use in this function); did you mean ‘EXIT_SUCCESS’?
  685 |     if (res != PAM_SUCCESS) {
      |                ^~~~~~~~~~~
      |                EXIT_SUCCESS
mod_sftp_pam.c:686:75: warning: format ‘%s’ expects argument of type ‘char *’, but argument 4 has type ‘int’ [-Wformat=]
  686 |       pr_trace_msg(trace_channel, 9, "PAM error setting PAM_DELETE_CRED: %s",
      |                                                                          ~^
      |                                                                           |
      |                                                                           char *
      |                                                                          %d
  687 |         pam_strerror(sftppam_pamh, res));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                    
      |         |
      |         int
mod_sftp_pam.c:691:11: warning: implicit declaration of function ‘pam_close_session’ [-Wimplicit-function-declaration]
  691 |     res = pam_close_session(sftppam_pamh, PAM_SILENT);
      |           ^~~~~~~~~~~~~~~~~
mod_sftp_pam.c:691:43: error: ‘PAM_SILENT’ undeclared (first use in this function); did you mean ‘CF_SILENT’?
  691 |     res = pam_close_session(sftppam_pamh, PAM_SILENT);
      |                                           ^~~~~~~~~~
      |                                           CF_SILENT
mod_sftp_pam.c:693:67: warning: format ‘%s’ expects argument of type ‘char *’, but argument 4 has type ‘int’ [-Wformat=]
  693 |       pr_trace_msg(trace_channel, 1, "error closing PAM session: %s",
      |                                                                  ~^
      |                                                                   |
      |                                                                   char *
      |                                                                  %d
  694 |         pam_strerror(sftppam_pamh, res));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                            
      |         |
      |         int
mod_sftp_pam.c:698:11: warning: implicit declaration of function ‘pam_end’; did you mean ‘va_end’? [-Wimplicit-function-declaration]
  698 |     res = pam_end(sftppam_pamh, PAM_SUCCESS);
      |           ^~~~~~~
      |           va_end
mod_sftp_pam.c:700:66: warning: format ‘%s’ expects argument of type ‘char *’, but argument 4 has type ‘int’ [-Wformat=]
  700 |       pr_trace_msg(trace_channel, 1, "error freeing PAM handle: %s",
      |                                                                 ~^
      |                                                                  |
      |                                                                  char *
      |                                                                 %d
  701 |         pam_strerror(sftppam_pamh, res));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                           
      |         |
      |         int
mod_sftp_pam.c: At top level:
mod_sftp_pam.c:271:30: error: storage size of ‘sftppam_conv’ isn’t known
  271 | static const struct pam_conv sftppam_conv = { &sftppam_converse, NULL };
      |                              ^~~~~~~~~~~~
make[1]: *** [Makefile:31: mod_sftp_pam.o] Error 1
make[1]: Leaving directory '/root/proftpd/modules'
make: *** [Makefile:57: modules] Error 2
```

I'm sure I'm doing something wrong, sorry.

**Comment by jlecour** [comment]
I've tried again with `./configure --with-modules=mod_sftp:mod_sftp_pam:mod_auth_pam` without a better result.

**Comment by jlecour** [comment]
I've installe `libpam0g-dev` and managed to compile without errors.

But now I can't get to enable the sftp module.

` ModulesPath` and `LoadModule` directives are not recognized:

```
# /usr/local/sbin/proftpd -c /usr/local/etc/proftpd.conf --configtest
Checking syntax of configuration file
2023-08-17 10:42:00,037 www00 proftpd[781762]: fatal: unknown configuration directive 'LoadModule' on line 8 of '/usr/local/etc/proftpd.conf'
```

```
# This is a basic ProFTPD configuration file (rename it to 
# 'proftpd.conf' for actual use.  It establishes a single server
# and a single anonymous login.  It assumes that you have a user/group
# "nobody" and "ftp" for normal operation and anon.

LoadModule mod_sftp.c
# LoadModule mod_sftp_pam.c


ServerName			"ProFTPD Default Installation"
ServerType			standalone
DefaultServer			on

# Port 21 is the standard FTP port.
#Port				21

# Don't use IPv6 support by default.
UseIPv6				off

# Umask 022 is a good standard umask to prevent new dirs and files
# from being group and world writable.
Umask				022

# To prevent DoS attacks, set the maximum number of child processes
# to 30.  If you need to allow more than 30 concurrent connections
# at once, simply increase this value.  Note that this ONLY works
# in standalone mode, in inetd mode you should use an inetd server
# that allows you to limit maximum number of processes per service
# (such as xinetd).
MaxInstances			30

# Set the user and group under which the server will run.
User				nobody
Group				nogroup

# To cause every FTP user to be "jailed" (chrooted) into their home
# directory, uncomment this line.
#DefaultRoot ~

# Normally, we want files to be overwriteable.
AllowOverwrite		on

# Bar use of SITE CHMOD by default
<Limit SITE_CHMOD>
  DenyAll
</Limit>

# A basic anonymous configuration, no upload directories.  If you do not
# want anonymous users, simply delete this entire <Anonymous> section.
<Anonymous ~ftp>
  User				ftp
  Group				ftp

  # We want clients to be able to login with "anonymous" as well as "ftp"
  UserAlias			anonymous ftp

  # Limit the maximum number of anonymous logins
  MaxClients			10

  # We want 'welcome.msg' displayed at login, and '.message' displayed
  # in each newly chdired directory.
  DisplayLogin			welcome.msg
  DisplayChdir			.message

  # Limit WRITE everywhere in the anonymous chroot
  <Limit WRITE>
    DenyAll
  </Limit>
</Anonymous>

#############################################################################

AuthUserFile      /etc/proftpd/vpasswd

# Connexion limits
MaxInstances      50
MaxClients        40
MaxClientsPerHost 20

# Network settings
PassivePorts  60000 61000
UseReverseDNS off

<IfModule mod_ident.c>
IdentLookups  off
</IfModule>

TimesGMT      off

<IfModule !mod_sftp.c>
Port 21
</IfModule>

# Local permissions
DefaultRoot       ~
Umask             137  027
RequireValidShell off
UseFtpUsers       off

# Allow RESUME (REST command)
AllowStoreRestart on

AuthOrder    mod_auth_file.c
AuthUserFile /etc/proftpd/vpasswd

<IfModule mod_sftp.c>
    SFTPEngine   on
    Port         2022
    #DefaultRoot  ~
    
    SFTPLog      /var/log/proftpd/sftp.log
    
    SFTPAuthMethods password publickey

    SFTPHostKey /etc/proftpd/ssh/ssh_host_ed25519_key
    SFTPHostKey /etc/proftpd/ssh/ssh_host_rsa_key
    SFTPHostKey /etc/proftpd/ssh/ssh_host_ecdsa_key
    
    SFTPAuthorizedUserKeys file:/etc/proftpd/sftp.passwd.keys/%u
    AuthUserFile /etc/proftpd/vpasswd
    
    RequireValidShell off

    # Enable compression
    # SFTPCompression delayed
</IfModule>
```

**Comment by jlecour** [comment]
I finally got to compile with mod_sftp.

When I start the service, I'm asked for a passphrase, but I'm pretty sure my host keys don't have any passphrase (generated with `ssh-keygen -f ./ssh_host_ed25519_key -t ed25519 -N '' -m PEM`).

**Comment by jlecour** [comment]
Elsewhere, I've been pointed at the way Debian is packaging ProFTPd, which is quite complicated : https://salsa.debian.org/debian-proftpd-team/proftpd/-/blob/master/debian/rules

I definitely want to try to reproduce my problem with a custom compiled ProFTPd, but if I couldn't, it wouldn't rule out a Debian-specific problem.

**Comment by Castaglia** [comment]
> Elsewhere, I've been pointed at the way Debian is packaging ProFTPd, which is quite complicated : https://salsa.debian.org/debian-proftpd-team/proftpd/-/blob/master/debian/rules

Indeed.  The very configurability of ProFTPD (dynamic _vs_ shared modules, compile-time _vs_ run-time settings and locations of files, libraries for modules, _etc_) can make it more complicated to build from source than one might expect.  Thanks for your persistence in this!

If it helps, the output from `proftpd -V` (as from your package installation) is meant to help reproduce all the build-time options (including list of modules, libraries, other settings), so that you can essentially copy/paste the `Built With` line, for cases like this.

> I definitely want to try to reproduce my problem with a custom compiled ProFTPd, but if I couldn't, it wouldn't rule out a Debian-specific problem.

That's a possibility as well.  We'll see what information we can find, and if it does appear to be specific to Debian somehow, we'll take it up with the ProFTPD package maintainers there.

I'm also still trying to reproduce this issue locally.  I'll take a closer look at versions of libraries (such as libsodium, OpenSSL) to see if any changes in those libraries might be related to the issues we're seeing here.

**Comment by Castaglia** [comment]
> > Could you send me the hostkey files you are using, via direct email?
> 
> I've generated a brand new set of host keys with `ssh-keygen` and the result is the same. If you still want that keys, I can share them and make some new ones again for me.

Yes, if you could share a set of nonworking host keys with me directly via email, for my local reproduction tests, that would be much appreciated!

**Comment by jlecour** [comment]
> Could you send me the hostkey files you are using, via direct email?

I've just done that.

**Comment by Castaglia** [comment]
I'm able to reproduce this issue locally now, thanks to the keys provided by @jlecour .

The cause _is_ the use of static buffer sizes within ProFTPD's `mod_sftp`, but it is _not_ in the location that I'd initially expected.  This issue is not specific to Cyberduck, or to the use of EC or Curve25519 algorithms; it's mostly caused by the sheer size of "KEXINIT" messages from both sides, _i.e._ the ever-growing list of supported algorithms by clients and servers.  These messages are used, per the protocol, as part of generating an SSH session identifier (called `H` ("hash") in the spec).  ProFTPD's buffers for computing these `H` values are not large enough in some cases -- such as this one.

I should have a fix shortly.

**Comment by Castaglia** [comment]
The fix has been merged to the `master` branch, and backported to the 1.3.8 branch.  Thanks!

**Comment by jlecour** [comment]
Thanks a lot.
I've opened an issue in Debian BTS hoping that this might be fixed in Debian 12 : https://bugs.debian.org/cgi-bin/%3Ca%20href=%22bugreport.cgi?bug=1051236%22%3E1051236%3C/a%3E

**Comment by Castaglia** [comment]
> Thanks a lot. I've opened an issue in Debian BTS hoping that this might be fixed in Debian 12 : https://bugs.debian.org/1051236

If not, I suspect that @hpreusse can help.
